aeneas/src/mach/GlobalRegAlloc.v3

```diff
@@ -345,14 +357,143 @@ class GlobalRegAlloc(regSet: MachRegSet, codegen: SsaMachGen) {
 					return result;
 				}
 			}
-			if (graph.nodes[node].spill) {
-				allocSpill(vreg);
-				return result;
-			}
-			return (false, null);
+			if (!graph.nodes[node].spill) return V3.fail("no colors available");
+			vreg.notSpillable = true;
+			allocSpill(vreg);
+			return (false, result.1);
 		}
 		return result;
 	}
+	def dupVReg(vreg: VReg, constraint: int) -> VReg {
+		var newVreg = codegen.dupVReg(vreg);
+		newVreg.notSpillable = true;
+		newVreg.reloadFrom = vreg;
+		if (constraint < regSet.physRegs) newVreg.hint = byte.view(constraint);
+		return newVreg;
+	}
+	def isOkDominator(vreg: VReg, newCommonDominator: SsaBlock) -> bool {
+		var node = graph.nodes[vreg.varNum];
+		if (newCommonDominator == node.commonDominator) return true;
+		if (newCommonDominator == node.defBlock) return false;
+		for (e in newCommonDominator.succs()) {
+			var succ = e.dest;
+			if (!liveness[succ.mark, vreg.varNum]) return false;
+		}
+		return true;
+	}
+	def pullBlockOutOfLoop(node: Node) -> SsaBlock {
+		var block = codegen.order[node.commonDominator.mark];
+		for (p = block; p != null; p = p.dom_parent) {
+			if (p.loop != null && p.dom_parent != null && p.dom_parent.block != node.defBlock) block = p.dom_parent;
+		}
+		return block.block;
+	}
+	// Try to insert reload in the common dominator.
+	def processSpillOnBlock(vreg: VReg, block: SsaBlock) {
+		var node = graph.nodes[vreg.varNum];
+		if (node.commonDominator == null) {
+			node.newVreg = codegen.dupVReg(vreg);
+			node.newVreg.spill = vreg.spill;
+			node.commonDominator = block;
+		} else {
+			var order = codegen.order;
+			var newCommonDominator = codegen.blocks.commonDominator(order[node.commonDominator.mark], order[block.mark]);
```

Comment:
I think it will be clearer to pull the computation of the new common dominator out into a method, rather than having `isOkDominator` and `pullBlockOutOfLoop` as separate routines.

Reply:
I put the computation and check for the new common dominator together. However, I keep `findDominatorOutOfLoop` as a separate function because choosing an upper common dominator earlier might cause it to fail the check of successor.